// stream-chat-react-native-core/lib/commonjs/components/ChannelList/hooks/listeners/useNewMessageNotification.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useNewMessageNotification = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _react = require("react");
var _ChatContext = require("../../../../contexts/chatContext/ChatContext");
var _utils = require("../../utils");
var _utils2 = require("../utils");
var useNewMessageNotification = exports.useNewMessageNotification = function useNewMessageNotification(_ref) {
  var onNewMessageNotification = _ref.onNewMessageNotification,
    options = _ref.options,
    setChannels = _ref.setChannels;
  var _useChatContext = (0, _ChatContext.useChatContext)(),
    client = _useChatContext.client;
  (0, _react.useEffect)(function () {
    var handleEvent = function () {
      var _ref2 = (0, _asyncToGenerator2["default"])(_regenerator["default"].mark(function _callee(event) {
        var _event$channel, _event$channel2, filters, sort, channel, considerArchivedChannels;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              if (!(typeof onNewMessageNotification === 'function')) {
                _context.next = 4;
                break;
              }
              onNewMessageNotification(setChannels, event, options);
              _context.next = 15;
              break;
            case 4:
              if (options) {
                _context.next = 6;
                break;
              }
              return _context.abrupt("return");
            case 6:
              filters = options.filters, sort = options.sort;
              if (!((_event$channel = event.channel) != null && _event$channel.id && (_event$channel2 = event.channel) != null && _event$channel2.type)) {
                _context.next = 15;
                break;
              }
              _context.next = 10;
              return (0, _utils.getChannel)({
                client: client,
                id: event.channel.id,
                type: event.channel.type
              });
            case 10:
              channel = _context.sent;
              considerArchivedChannels = filters && filters.archived === false;
              if (!((0, _utils2.isChannelArchived)(channel) && considerArchivedChannels)) {
                _context.next = 14;
                break;
              }
              return _context.abrupt("return");
            case 14:
              setChannels(function (channels) {
                return channels ? (0, _utils.moveChannelUp)({
                  channels: channels,
                  channelToMove: channel,
                  sort: sort
                }) : channels;
              });
            case 15:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      return function handleEvent(_x) {
        return _ref2.apply(this, arguments);
      };
    }();
    var listener = client == null ? void 0 : client.on('notification.message_new', handleEvent);
    return function () {
      return listener == null ? void 0 : listener.unsubscribe();
    };
  }, []);
};
