// stream-chat-react-native-core/lib/module/components/ChannelList/hooks/listeners/useAddedToChannelNotification.js
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useAddedToChannelNotification = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _react = require("react");
var _uniqBy = _interopRequireDefault(require("lodash/uniqBy"));
var _ChatContext = require("../../../../contexts/chatContext/ChatContext");
var _utils = require("../../utils");
var _utils2 = require("../utils");
var useAddedToChannelNotification = exports.useAddedToChannelNotification = function useAddedToChannelNotification(_ref) {
  var onAddedToChannel = _ref.onAddedToChannel,
    options = _ref.options,
    setChannels = _ref.setChannels;
  var _useChatContext = (0, _ChatContext.useChatContext)(),
    client = _useChatContext.client;
  (0, _react.useEffect)(function () {
    var handleEvent = function () {
      var _ref2 = (0, _asyncToGenerator2["default"])(_regenerator["default"].mark(function _callee(event) {
        var _event$channel, _event$channel2, sort, channel, pinnedAtSort;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              if (!(typeof onAddedToChannel === 'function')) {
                _context.next = 4;
                break;
              }
              onAddedToChannel(setChannels, event, options);
              _context.next = 13;
              break;
            case 4:
              if (options) {
                _context.next = 6;
                break;
              }
              return _context.abrupt("return");
            case 6:
              sort = options.sort;
              if (!((_event$channel = event.channel) != null && _event$channel.id && (_event$channel2 = event.channel) != null && _event$channel2.type)) {
                _context.next = 13;
                break;
              }
              _context.next = 10;
              return (0, _utils.getChannel)({
                client: client,
                id: event.channel.id,
                type: event.channel.type
              });
            case 10:
              channel = _context.sent;
              pinnedAtSort = (0, _utils2.findPinnedAtSortOrder)({
                sort: sort
              });
              setChannels(function (channels) {
                if (!channels) {
                  return channels;
                }
                var lastPinnedChannelIndex = null;
                var newChannels = (0, _toConsumableArray2["default"])(channels);
                if (pinnedAtSort === 1 || pinnedAtSort === -1) {
                  lastPinnedChannelIndex = (0, _utils2.findLastPinnedChannelIndex)({
                    channels: newChannels
                  });
                  var newTargetChannelIndex = typeof lastPinnedChannelIndex === 'number' ? lastPinnedChannelIndex + 1 : 0;
                  newChannels.splice(newTargetChannelIndex, 0, channel);
                  return newChannels;
                }
                return (0, _uniqBy["default"])([channel].concat((0, _toConsumableArray2["default"])(channels)), 'cid');
              });
            case 13:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      return function handleEvent(_x) {
        return _ref2.apply(this, arguments);
      };
    }();
    var listener = client == null ? void 0 : client.on('notification.added_to_channel', handleEvent);
    return function () {
      return listener == null ? void 0 : listener.unsubscribe();
    };
  }, []);
};
